DOC: 2.0 release highlights and compat notes changes (#25937)
numpy/numpy: doc/source/release/2.0.0-notes.rst

NumPy 2.0.0 Release Notes
=========================

NumPy 2.0.0 is the first major release since 2006. It is the result of X months
of development since the last feature release by Y contributors, and contains a
large amount of exciting new features as well as a large amount of changes to
both the Python and C APIs.

> 

Author: Comment by @charris: start with a list of links to other docs that are relevant (migration guide, etc)

Highlights of this release include:

- New features:

  - A new variable-length string dtype, `~numpy.dtypes.StringDType` and a new
    `numpy.strings` namespace with performant ufuncs ...
  - Support for ``float32`` and ``longdouble`` in all `numpy.fft` functions,
  - Support for the array API standard in the main ``numpy`` namespace

- Performance improvements:

>   - Sorting functions (`sort`, `argsort`, `partition`)

Reviewer: ```suggestion
  - Sorting functions (`sort`, `argsort`, `partition`, `argpartition`)
```

Author: Thanks @r-devulap. Useful comments, all incorporated now.

>     have been accelerated through use of SIMD kernels and Highway, and may

Reviewer: ```suggestion
    have been accelerated through the use of Intel x86-simd-sort and Google Highway libraries, and may
```

NumPy 2.0.0 Release Notes
=========================

NumPy 2.0.0 is the first major release since 2006. It is the result of X months
of development since the last feature release by Y contributors, and contains a
large amount of exciting new features as well as a large amount of changes to

> both the Python and C APIs.

Reviewer 2: I am wondering if we can expand this section in some free-form to include:

Beyond general improvements and new features the major changes for Python users are Python API changes which for the most part are removals and cleanups but may require code adaptations (NEP 56) and the changes in promotion (NEP 50) which fix many user surprises but can lead to different precision results.

I.e. these are the most relevant (to end users) changes which couldn't happen in a 1.x release.  NEP 50 feels a bit lost below considering that it has quite a bit of potential for breaking things.
(The migration guide makes that clear I think, so I guess if you don't like listing anything here, you could also just stress that as a major release 2.0 had some changes which could affect users more, please see there there...)

Author: good idea, added something along these lines.

Key documents related to how to adapt to changes in NumPy 2.0, in addition to
these release notes, include:

- The :ref:`numpy-2-migration-guide`
- The :ref:`NumPy 2.0-specific advice <numpy-2-abi-handling>` in
  :ref:`for-downstream-package-authors`

Highlights
==========

Highlights of this release include:

- New features:

  - A new variable-length string dtype, `~numpy.dtypes.StringDType` and a new
    `numpy.strings` namespace with performant ufuncs ...
  - Support for ``float32`` and ``longdouble`` in all `numpy.fft` functions,
  - Support for the array API standard in the main ``numpy`` namespace

- Performance improvements:

  - Sorting functions (`sort`, `argsort`, `partition`, `argpartition`)
    have been accelerated through the use of the Intel x86-simd-sort and Google
    Highway libraries, and may see large (hardware-specific) speedups,
  - macOS Accelerate support and binary wheels for macOS >=14, with significant
    performance improvements for linear algebra operations on macOS

>   - `numpy.char` fixed-length string dtypes have been accelerated, ...

Reviewer 3: ```suggestion
  - `numpy.char` string operations have been accelerated by implementing
    ufuncs that also support `~numpy.dtypes.StringDType` in addition to the
    the fixed-length string dtypes.
```

  - A new tracing and introspection API to determine which hardware-specific
    kernels are available and will be dispatched to ... see
    `~numpy.lib.introspect.opt_func_info`

- Python API improvements:

  - A clear split between public and private API, with a new
    :ref:`module structure <module-structure>`, and each public function now
    available in a single place,
  - Many removals of non-recommended functions and aliases. This should make
    it easier to learn and use NumPy. The number of objects in the main
    namespace decreased by ~10% and in ``numpy.lib`` by ~80%,
  - :ref:`Canonical dtype names <canonical-python-and-c-types>` and a new
    `~numpy.isdtype` introspection function,

- C API improvements:

  - A new public C API for creating custom dtypes
  - Many outdated functions and macros removed, and private internals hidden to
    ease future extensibility,

>   - New, easier to use, initialization functions,

Reviewer 3: Is it standard to include trailing commas in list items?

Author: It's common; there's no style guide with this level of detail. This doc requires a full copy-edit later on, so I will resolve this for now.

- Improved behavior:

>   - Improvements to type promotion behavior (NEP 50) ...

Reviewer 2: ```suggestion
  - Improvements to type promotion behavior was changed by adopting `NEP 50<NEP50>`_.
    This makes fixes many user surprises about promotions which previously often
    dependet on the values.  Please see the NEP and the :ref:`numpy-2-migration-guide`
    for details as this change e.g. can lead to lower precision results.
```

  - The default integer type on Windows is now ``int64`` rather than ``int32``,
    matching the behavior on other platform,
  - The maximum number of array dimensions is changed from 32 to 64

- Documentation:

  - The reference guide navigation was signficantly improved, and there is now
    documentation on NumPy's :ref:`module structure <module-structure>`,
  - The :ref:`building from source <building-from-source>` documentation was
    completely rewritten,

Furthermore there are many changes to NumPy internals, including continuing to
migrate code from C to C++, that will make it easier to improve and maintain
NumPy in the future.

The "no free lunch" theorem dictates that there is a price to pay for all these
API and behavior improvements and better future extensibility. This price is:

1. Backwards compatibility - clear errors in most cases; see the Deprecation
   and Compatibility sections below, and the :ref:`numpy-2-migration-guide`
   (note: there is a ``ruff`` mode to auto-fix many things in Python code)
2. ABI change - impact on downstream libraries and binaries. See
   :ref:`numpy-2-abi-handling`.
   TODO: main messages.

> 

Author: I would not mention that now. Let's first see how it goes. Numba is usually 3-6 months late. If all of scipy/matplotlib/pandas/scikits have compatible releases up before we do 2.0.0 final, then I don't think we need to say this.

> 

Reviewer 3: Delete second empty line.

```suggestion

```

The Python versions supported by this release are 3.9-3.12.

> 

Reviewer 3: Delete third empty line
```suggestion

```

(`gh-25336 <https://github.com/numpy/numpy/pull/25336>`__)

Minor changes in behavior of sorting functions
----------------------------------------------

Due to algorithmic changes and use of SIMD code, sorting functions with methods
that aren't stable may return slightly different results in 2.0.0 compared to

> 1.26.x. This includes the default method of `~numpy.sort` and `~numpy.argsort`.

Reviewer: ```suggestion
1.26.x. This includes the default method of `~numpy.argsort` and `~numpy.argpartition`.
```

The unstableness in quicksort shows up in the `arg` methods only.
